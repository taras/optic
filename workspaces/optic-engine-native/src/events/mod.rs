#![allow(dead_code)]

use chrono::{DateTime, Local, TimeZone, Utc};
pub use cqrs_core::{AggregateEvent, Event};
use serde::{Deserialize, Serialize, Serializer};
use serde_json;
use std::fs;
use std::io;
use std::path::Path;

pub mod endpoint;
pub mod http_interaction;
pub mod rfc;
pub mod shape;
pub mod spec_chunk;

pub use endpoint::EndpointEvent;
pub use http_interaction::HttpInteraction;
pub use rfc::RfcEvent;
pub use shape::ShapeEvent;
pub use spec_chunk::SpecChunkEvent;

use crate::CommandContext;

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(untagged)]
pub enum SpecEvent {
  EndpointEvent(endpoint::EndpointEvent),
  RfcEvent(rfc::RfcEvent),
  ShapeEvent(shape::ShapeEvent),
}

impl Event for SpecEvent {
  fn event_type(&self) -> &'static str {
    match self {
      SpecEvent::EndpointEvent(evt) => evt.event_type(),
      SpecEvent::RfcEvent(evt) => evt.event_type(),
      SpecEvent::ShapeEvent(evt) => evt.event_type(),
    }
  }
}

impl WithEventContext for SpecEvent {
  fn with_event_context(&mut self, event_context: EventContext) {
    match self {
      SpecEvent::EndpointEvent(evt) => evt.with_event_context(event_context),
      SpecEvent::RfcEvent(evt) => evt.with_event_context(event_context),
      SpecEvent::ShapeEvent(evt) => evt.with_event_context(event_context),
    };
  }
}

impl SpecEvent {
  pub fn from_file(filename: impl AsRef<Path>) -> Result<Vec<SpecEvent>, EventLoadingError> {
    let file_contents = fs::read_to_string(filename)?;

    let events: Vec<SpecEvent> = serde_json::from_str(&file_contents)?;

    Ok(events)
  }
}

impl From<EndpointEvent> for SpecEvent {
  fn from(endpoint_event: EndpointEvent) -> Self {
    Self::EndpointEvent(endpoint_event)
  }
}

impl From<RfcEvent> for SpecEvent {
  fn from(rfc_event: RfcEvent) -> Self {
    Self::RfcEvent(rfc_event)
  }
}

impl From<ShapeEvent> for SpecEvent {
  fn from(shape_event: ShapeEvent) -> Self {
    Self::ShapeEvent(shape_event)
  }
}

#[derive(Debug)]
pub enum EventLoadingError {
  Avro(avro_rs::Error),
  Io(io::Error),
  Json(serde_json::Error),
}

impl From<io::Error> for EventLoadingError {
  fn from(err: io::Error) -> EventLoadingError {
    EventLoadingError::Io(err)
  }
}

impl From<serde_json::Error> for EventLoadingError {
  fn from(err: serde_json::Error) -> EventLoadingError {
    use serde_json::error::Category;
    match err.classify() {
      Category::Io => EventLoadingError::Io(err.into()),
      Category::Syntax | Category::Data | Category::Eof => EventLoadingError::Json(err),
    }
  }
}

impl From<avro_rs::Error> for EventLoadingError {
  fn from(err: avro_rs::Error) -> EventLoadingError {
    EventLoadingError::Avro(err)
  }
}

// EventContext
// ------------

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct EventContext {
  pub client_id: String,
  pub client_session_id: String,
  pub client_command_batch_id: String,

  #[serde(serialize_with = "created_at_to_local_timezone")]
  pub created_at: String,
}

fn created_at_to_local_timezone<S>(created_at: &String, serializer: S) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  if let Ok(parsed_date) = DateTime::parse_from_rfc3339(created_at) {
    let local = parsed_date.with_timezone(&Local);
    serializer.serialize_str(&local.to_rfc3339())
  } else {
    serializer.serialize_str(&created_at)
  }
}

impl From<CommandContext> for EventContext {
  fn from(command_context: CommandContext) -> Self {
    Self {
      client_id: command_context.client_id,
      client_session_id: command_context.client_session_id,
      client_command_batch_id: command_context.client_command_batch_id,
      created_at: command_context.created_at.to_rfc3339(),
    }
  }
}

pub trait WithEventContext {
  fn with_event_context(&mut self, event_context: EventContext);
}
