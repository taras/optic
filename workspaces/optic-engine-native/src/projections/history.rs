use cqrs_core::{Aggregate, AggregateEvent};
use petgraph::visit;
use petgraph::{
  graph::{Graph, NodeIndex},
  Directed,
};
use std::collections::HashMap;

use crate::events::RfcEvent;

pub const ROOT_COMMIT_ID: &str = "root";

pub type NodeId = String;
pub type CommitId = String;

#[derive(Debug)]
pub enum Node {
  BatchCommit(BatchCommitNode),
}

#[derive(Debug)]
pub struct BatchCommitNode(CommitId, BatchCommitDescriptor);

#[derive(Debug)]
pub struct BatchCommitDescriptor {
  commit_message: String,
  pub is_complete: bool,
}

#[derive(Debug)]
pub enum Edge {
  IsParentOf,
}

#[derive(Debug)]
pub struct HistoryProjection {
  pub graph: Graph<Node, Edge>,
  pub node_id_to_index: HashMap<NodeId, NodeIndex>,
}

impl Default for HistoryProjection {
  fn default() -> Self {
    let mut graph = Graph::new();
    let mut node_id_to_index = HashMap::new();

    let root_commit_id = CommitId::from(ROOT_COMMIT_ID);
    let root_node = Node::BatchCommit(BatchCommitNode(
      root_commit_id.clone(),
      BatchCommitDescriptor {
        commit_message: String::from("Initial commit"),
        is_complete: false,
      },
    ));
    let root_node_index = graph.add_node(root_node);
    node_id_to_index.insert(root_commit_id, root_node_index);

    HistoryProjection {
      graph,
      node_id_to_index,
    }
  }
}

impl HistoryProjection {
  fn with_batch_commit_start(
    &mut self,
    batch_id: CommitId,
    parent_batch_id: Option<CommitId>,
    commit_message: String,
  ) {
    let parent_batch_id = parent_batch_id.unwrap_or_else(|| {
      self
        .get_commit_id(&self.find_last_batch_commit_index().unwrap())
        .unwrap()
        .clone()
    });

    if parent_batch_id == ROOT_COMMIT_ID {
      self.with_batch_commit_end(CommitId::from(ROOT_COMMIT_ID));
    }

    let node = Node::BatchCommit(BatchCommitNode(
      batch_id.clone(),
      BatchCommitDescriptor {
        commit_message,
        is_complete: false,
      },
    ));
    let node_index = self.graph.add_node(node);
    self.node_id_to_index.insert(batch_id.clone(), node_index);

    let parent_node_index = *self
      .node_id_to_index
      .get(&parent_batch_id)
      .expect(&*format!(
        "expected parent commit to have a corresponding node {:?}",
        parent_batch_id
      ));

    self
      .graph
      .add_edge(parent_node_index, node_index, Edge::IsParentOf);
  }

  fn with_batch_commit_end(&mut self, batch_id: CommitId) {
    let commit_node_index_option = self.get_batch_commit_node_index(&batch_id);
    if let Some(commit_node_index) = commit_node_index_option {
      let commit_node_index = *commit_node_index;
      let commit_node_descriptor = match self.graph.node_weight_mut(commit_node_index).unwrap() {
        Node::BatchCommit(node_descriptor) => &mut node_descriptor.1,
      };

      commit_node_descriptor.is_complete = true;
    }
  }

  #[allow(irrefutable_let_patterns)]
  pub fn get_batch_commit_node_index(&self, batch_id: &CommitId) -> Option<&NodeIndex> {
    let node_index = self.node_id_to_index.get(batch_id)?;
    let node = self.graph.node_weight(*node_index)?;
    if let &Node::BatchCommit(_) = node {
      Some(node_index)
    } else {
      None
    }
  }

  #[allow(irrefutable_let_patterns)]
  pub fn get_batch_commit_node(&self, batch_id: &CommitId) -> Option<&Node> {
    let node_index = self.node_id_to_index.get(batch_id)?;
    let node = self.graph.node_weight(*node_index)?;
    if let &Node::BatchCommit(_) = node {
      Some(node)
    } else {
      None
    }
  }

  #[allow(irrefutable_let_patterns)]
  pub fn get_commit_id(&self, node_index: &NodeIndex) -> Option<&CommitId> {
    let node = self.graph.node_weight(*node_index)?;
    if let &Node::BatchCommit(BatchCommitNode(commit_id, _)) = &node {
      Some(commit_id)
    } else {
      None
    }
  }

  #[allow(unreachable_patterns)]
  pub fn get_batch_commit_descriptor(&self, batch_id: &CommitId) -> Option<&BatchCommitDescriptor> {
    let node = self.get_batch_commit_node(batch_id)?;
    match node {
      Node::BatchCommit(BatchCommitNode(_, descriptor)) => Some(descriptor),
      _ => None,
    }
  }

  // pub fn get_batch_commits_iter(
  //   &self,
  //   batch_id: &CommitId,
  // ) -> Option<impl Iterator<Item = &CommitId>> {
  //   let commit_node_index = *self.get_batch_commit_node_index(batch_id)?;

  //   let commit_graph = NodeFiltered::from_fn(&self.graph, |node_index| {
  //     matches!(
  //       self.graph.node_weight(node_index),
  //       Some(Node::BatchCommit(_))
  //     )
  //   });

  //   let commit_log = EdgeFiltered::from_fn(&commit_graph, |edge| {
  //     matches!(edge.weight(), Edge::IsChildOf)
  //       && self
  //         .graph
  //         .edges_directed(edge.target(), petgraph::Direction::Incoming)
  //         .into_iter()
  //         .count()
  //         == 1
  //   });

  //   let dfs = Dfs::new(&commit_log, commit_node_index)
  //     .iter()
  //     .map(|node_index| &commit_log.node_weight(node_index));
  //   Some(std::iter::empty())
  //

  pub fn find_last_batch_commit_index(&self) -> Option<NodeIndex> {
    let root_node_index = *self.get_batch_commit_node_index(&CommitId::from(ROOT_COMMIT_ID))?;
    let commit_graph = visit::NodeFiltered::from_fn(&self.graph, |node_index| {
      matches!(
        self.graph.node_weight(node_index),
        Some(Node::BatchCommit(_))
      )
    });

    // @GOTCHA: this assumes there's the commit graph is a commit log, with no branches
    let mut leaf_node_index = Some(root_node_index);
    visit::depth_first_search(&self.graph, Some(root_node_index), |event| {
      if let visit::DfsEvent::TreeEdge(parent_node_index, child_node_index) = event {
        leaf_node_index.replace(child_node_index);
      }

      if let visit::DfsEvent::BackEdge(_, _) = event {
        visit::Control::Break(())
      } else {
        visit::Control::Continue
      }
    });

    leaf_node_index
  }
}

impl Aggregate for HistoryProjection {
  fn aggregate_type() -> &'static str {
    "history_projection"
  }
}

// Events
// ------

impl AggregateEvent<HistoryProjection> for RfcEvent {
  fn apply_to(self, projection: &mut HistoryProjection) {
    match self {
      RfcEvent::BatchCommitStarted(e) => {
        projection.with_batch_commit_start(e.batch_id, e.parent_id, e.commit_message);
      }
      RfcEvent::BatchCommitEnded(e) => {
        projection.with_batch_commit_end(e.batch_id);
      }
      // explicitly ignore other rfc events, so we don't silently swallow newly added events
      RfcEvent::APINamed(_) | RfcEvent::ContributionAdded(_) | RfcEvent::GitStateSet(_) => {}
    }
  }
}

impl<I> From<I> for HistoryProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = HistoryProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}
