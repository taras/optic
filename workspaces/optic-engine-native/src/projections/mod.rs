pub mod conflicts;
pub mod contributions;
pub mod endpoint;
pub mod history;
pub mod learners;
pub mod shape;
pub mod spec_events;
pub mod spectacle;

pub use conflicts::ConflictsProjection;
pub use contributions::ContributionsProjection;
pub use endpoint::{EndpointProjection, ResponseBodyDescriptor};
pub use history::{CommitId, HistoryProjection};
pub use learners::{
  shape_diff_affordances::LearnedShapeDiffAffordancesProjection,
  undocumented_bodies::LearnedUndocumentedBodiesProjection,
};
pub use shape::ShapeProjection;
pub use spec_events::{SpecAssemblerError, SpecAssemblerProjection};
pub use spectacle::endpoints::EndpointsProjection;

use crate::events::{EndpointEvent, RfcEvent, ShapeEvent, SpecEvent};
use cqrs_core::{Aggregate, AggregateCommand, AggregateEvent, CommandError};
use std::error::Error;

#[derive(Debug)]
pub struct SpecProjection {
  endpoint: endpoint::EndpointProjection,
  history: history::HistoryProjection,
  shape: shape::ShapeProjection,
  conflicts: conflicts::ConflictsProjection,
  spectacle_endpoints: spectacle::endpoints::EndpointsProjection,
  contributions: contributions::ContributionsProjection,
}

impl Default for SpecProjection {
  fn default() -> Self {
    Self {
      endpoint: EndpointProjection::default(),
      history: HistoryProjection::default(),
      shape: ShapeProjection::default(),
      conflicts: ConflictsProjection::default(),
      spectacle_endpoints: EndpointsProjection::default(),
      contributions: ContributionsProjection::default(),
    }
  }
}

impl SpecProjection {
  pub fn endpoint(&self) -> &EndpointProjection {
    &self.endpoint
  }

  pub fn history(&self) -> &HistoryProjection {
    &self.history
  }

  pub fn shape(&self) -> &ShapeProjection {
    &self.shape
  }
  pub fn conflicts(&self) -> &ConflictsProjection {
    &self.conflicts
  }
  pub fn contributions(&self) -> &ContributionsProjection {
    &self.contributions
  }
  pub fn spectacle_endpoints(&self) -> &EndpointsProjection {
    &self.spectacle_endpoints
  }
  pub fn shapes_serializable(&self) -> crate::projections::shape::SerializableGraph {
    self.shape.to_serializable_graph()
  }
  pub fn spectacle_endpoints_serializable(
    &self,
  ) -> crate::projections::spectacle::endpoints::SerializableGraph {
    self.spectacle_endpoints.to_serializable_graph()
  }
}

impl Aggregate for SpecProjection {
  fn aggregate_type() -> &'static str {
    "spec_projection"
  }
}

// Events
// ------

impl AggregateEvent<SpecProjection> for SpecEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    match self {
      SpecEvent::EndpointEvent(event) => {
        projection.endpoint.apply(event.clone());
        projection.spectacle_endpoints.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::ShapeEvent(event) => {
        projection.shape.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::RfcEvent(event) => {
        projection.history.apply(event.clone());
        projection.shape.apply(event.clone());
        projection.spectacle_endpoints.apply(event.clone());
        projection.contributions.apply(event.clone());
      }
    }
  }
}

impl AggregateEvent<HistoryProjection> for SpecEvent {
  fn apply_to(self, projection: &mut HistoryProjection) {
    if let SpecEvent::RfcEvent(event) = self {
      event.apply_to(projection);
    }
  }
}

impl<I> From<I> for SpecProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = SpecProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}

// Convient application of spec event variants
// -------------------------------------------
impl AggregateEvent<SpecProjection> for ShapeEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::ShapeEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for RfcEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::RfcEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for EndpointEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::EndpointEvent(self))
  }
}
