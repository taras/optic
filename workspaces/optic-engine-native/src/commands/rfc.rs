use super::{CommandContext, SpecCommand, SpecCommandError};
use crate::events::rfc as rfc_events;
use crate::events::RfcEvent;
use crate::projections::{CommitId, HistoryProjection};
use crate::queries::history::HistoryQueries;
use cqrs_core::AggregateCommand;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Debug, Clone, Serialize)]
pub enum RfcCommand {
  AddContribution(AddContribution),
  SetAPIName(SetAPIName),
  SetGitState(SetGitState),
  MarkSetupStageComplete(MarkSetupStageComplete),

  #[serde(skip)]
  StartBatchCommit(StartBatchCommit),
  #[serde(skip)]
  EndBatchCommit(EndBatchCommit),
  #[serde(skip)]
  AppendBatch(AppendBatch),
}

impl RfcCommand {
  pub fn append_batch_commit(batch_id: String, commit_message: String) -> Self {
    Self::AppendBatch(AppendBatch {
      batch_id,
      commit_message,
    })
  }

  pub fn start_batch_commit(batch_id: String, parent_id: String, commit_message: String) -> Self {
    Self::StartBatchCommit(StartBatchCommit {
      batch_id,
      parent_id,
      commit_message,
    })
  }

  pub fn end_batch_commit(batch_id: String) -> Self {
    Self::EndBatchCommit(EndBatchCommit { batch_id })
  }

  pub fn add_contribution(id: String, key: String, value: String) -> Self {
    Self::AddContribution(AddContribution { id, key, value })
  }
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AddContribution {
  pub id: String,
  pub key: String,
  pub value: String,
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SetAPIName {
  new_name: String,
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SetGitState {
  commit_id: String,
  branch_name: String,
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MarkSetupStageComplete {
  step: String,
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StartBatchCommit {
  pub batch_id: String,
  pub parent_id: String,
  pub commit_message: String,
}

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndBatchCommit {
  pub batch_id: String,
}

#[derive(Debug, Clone)]
pub struct AppendBatch {
  pub batch_id: String,
  pub commit_message: String,
}

// ComandHandling
// --------------

impl AggregateCommand<HistoryProjection> for RfcCommand {
  type Error = SpecCommandError;
  type Event = RfcEvent;
  type Events = Vec<RfcEvent>;

  fn execute_on(self, history_projection: &HistoryProjection) -> Result<Self::Events, Self::Error> {
    let validation = CommandValidationQueries::from((history_projection, &self));

    let events = match self {
      RfcCommand::AppendBatch(command) => {
        let parent_batch_id = {
          let history_queries = HistoryQueries::from(history_projection);
          history_queries.resolve_latest_batch_commit_id()
        };

        vec![RfcEvent::from(RfcCommand::start_batch_commit(
          command.batch_id.clone(),
          parent_batch_id,
          command.commit_message.clone(),
        ))]
      }

      RfcCommand::EndBatchCommit(command) => {
        validation.require(
          validation.batch_commit_id_exists(&command.batch_id),
          "batch commit must exist to end batch commit",
        )?;
        validation.require(
          !validation.batch_commit_completed(&command.batch_id),
          "batch commit can not have ended yet to end batch commit",
        )?;

        vec![RfcEvent::from(rfc_events::BatchCommitEnded::from(command))]
      }

      RfcCommand::AddContribution(command) => {
        vec![RfcEvent::from(rfc_events::ContributionAdded::from(command))]
      }

      _ => Err(SpecCommandError::Unimplemented(
        "rfc command not implemented for rfc projection",
        SpecCommand::RfcCommand(self),
      ))?,
    };

    Ok(events)
  }
}

struct CommandValidationQueries<'a> {
  command_description: String,
  history_projection: &'a HistoryProjection,
}

impl<'a> CommandValidationQueries<'a> {
  fn require(&self, condition: bool, msg: &'static str) -> Result<(), SpecCommandError> {
    if condition {
      Ok(())
    } else {
      Err(SpecCommandError::Validation(format!(
        "Command failed validation: {}, {:?}",
        msg, self.command_description
      )))
    }
  }

  pub fn batch_commit_id_exists(&self, commit_id: &CommitId) -> bool {
    self
      .history_projection
      .get_batch_commit_node_index(commit_id)
      .is_some()
  }

  pub fn batch_commit_completed(&self, commit_id: &CommitId) -> bool {
    let descriptor = self
      .history_projection
      .get_batch_commit_descriptor(commit_id);
    descriptor.is_some() && descriptor.unwrap().is_complete
  }
}

impl<'a> From<(&'a HistoryProjection, &RfcCommand)> for CommandValidationQueries<'a> {
  fn from((history_projection, rfc_command): (&'a HistoryProjection, &RfcCommand)) -> Self {
    Self {
      command_description: format!("{:?}", rfc_command),
      history_projection,
    }
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::events::SpecEvent;
  use cqrs_core::Aggregate;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  pub fn can_handle_append_batch_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
      {"BatchCommitStarted": {"batchId": "batch_1", "commitMessage": "initial commit"}},
      {"BatchCommitEnded": {"batchId": "batch_1" }}
    ]))
    .expect("initial events should be valid spec events");

    let mut projection = HistoryProjection::from(initial_events);

    let command: RfcCommand =
      RfcCommand::append_batch_commit(String::from("test-batch-1"), String::from("second commit"));

    let new_events = projection
      .execute(command)
      .expect("new command should be applicable to initial projection");
    assert_eq!(new_events.len(), 1);
    // TODO: snapshot the results instead, once we figure out how insta redactions can control
    // for the automatically generated batch id.
    assert!(matches!(new_events[0], RfcEvent::BatchCommitStarted(_)));
    let parent_id = match &new_events[0] {
      RfcEvent::BatchCommitStarted(batch_start) => batch_start.parent_id.clone().unwrap(),
      _ => unreachable!(),
    };
    assert_eq!(parent_id, "batch_1");
    // assert_debug_snapshot!("can_handle_append_batch_command__new_events", new_events);

    for event in new_events {
      projection.apply(event);
    }
  }

  #[test]
  pub fn can_handle_end_batch_commit_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
      {"BatchCommitStarted": {"batchId": "batch_1","parentId": "root", "commitMessage": "initial commit"}},
      {"BatchCommitEnded": {"batchId": "batch_1" }},
      {"BatchCommitStarted": {"batchId": "batch_2","parentId": "batch_1", "commitMessage": "second commit"}},
    ]))
    .expect("initial events should be valid spec events");

    let mut projection = HistoryProjection::from(initial_events);

    let command: RfcCommand = RfcCommand::end_batch_commit(String::from("batch_2"));

    let new_events = projection
      .execute(command)
      .expect("new command should be applicable to initial projection");
    assert_eq!(new_events.len(), 1);
    assert_debug_snapshot!(
      "can_handle_end_batch_commit_command__new_events",
      new_events
    );

    let unexisting_batch: RfcCommand = RfcCommand::end_batch_commit(String::from("not-a-batch"));
    let unexisting_batch_result = projection.execute(unexisting_batch);
    assert!(unexisting_batch_result.is_err());
    assert_debug_snapshot!(
      "can_handle_end_batch_commit_command__unexisting_batch_result",
      unexisting_batch_result.unwrap_err()
    );

    let already_ended_batch: RfcCommand = RfcCommand::end_batch_commit(String::from("batch_1"));
    let already_ended_batch_result = projection.execute(already_ended_batch);
    assert!(already_ended_batch_result.is_err());
    assert_debug_snapshot!(
      "can_handle_end_batch_commit_command__already_ended_batch_result",
      already_ended_batch_result.unwrap_err()
    );

    for event in new_events {
      projection.apply(event);
    }
  }

  #[test]
  pub fn can_handle_contribution_commands() {
    let initial_events: Vec<SpecEvent> =
      serde_json::from_value(json!([])).expect("initial events should be valid spec events");

    let projection = HistoryProjection::from(initial_events);

    let command: RfcCommand = RfcCommand::add_contribution(
      String::from("path123.method"),
      String::from("purpose"),
      String::from("Name of Endpoint"),
    );

    let new_events = projection
      .execute(command)
      .expect("new contributions should be applicable projection");
    assert_eq!(new_events.len(), 1);
    assert_debug_snapshot!(
      "can_handle_add_contribution_command__new_events",
      new_events
    );
  }
}
