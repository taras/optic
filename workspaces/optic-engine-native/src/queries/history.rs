use crate::projections::history::{CommitId, HistoryProjection, ROOT_COMMIT_ID};

pub struct HistoryQueries<'a> {
  history_projection: &'a HistoryProjection,
}

impl<'a> From<&'a HistoryProjection> for HistoryQueries<'a> {
  fn from(history_projection: &'a HistoryProjection) -> Self {
    HistoryQueries { history_projection }
  }
}

impl<'a> HistoryQueries<'a> {
  pub fn resolve_latest_batch_commit_id(&self) -> CommitId {
    let commit_id = self
      .history_projection
      .find_last_batch_commit_index()
      .map(|node_index| {
        self
          .history_projection
          .get_commit_id(&node_index)
          .map(|id| id.clone())
      })
      .flatten();

    commit_id.unwrap_or_else(|| CommitId::from(ROOT_COMMIT_ID))
  }
}
