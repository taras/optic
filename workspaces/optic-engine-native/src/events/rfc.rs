use super::{EventContext, WithEventContext};
use crate::commands::rfc as rfc_commands;
use crate::commands::RfcCommand;
use cqrs_core::Event;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// RFC Events
// -----------
#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
pub enum RfcEvent {
  ContributionAdded(ContributionAdded),
  APINamed(APINamed),
  GitStateSet(GitStateSet),
  BatchCommitStarted(BatchCommitStarted),
  BatchCommitEnded(BatchCommitEnded),
}

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ContributionAdded {
  pub id: String,
  pub key: String,
  pub value: String,
  event_context: Option<EventContext>,
}

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct APINamed {
  name: String,
  event_context: Option<EventContext>,
}

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitStateSet {
  branch_name: String,
  commit_id: String,
  event_context: Option<EventContext>,
}

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct BatchCommitStarted {
  pub batch_id: String,
  pub commit_message: String,
  pub(crate) event_context: Option<EventContext>,

  #[serde(skip_serializing_if = "Option::is_none")]
  pub parent_id: Option<String>,
}

#[derive(Deserialize, Debug, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct BatchCommitEnded {
  pub batch_id: String,
  event_context: Option<EventContext>,
}

impl Event for RfcEvent {
  fn event_type(&self) -> &'static str {
    match self {
      RfcEvent::ContributionAdded(evt) => evt.event_type(),
      RfcEvent::APINamed(evt) => evt.event_type(),
      RfcEvent::GitStateSet(evt) => evt.event_type(),
      RfcEvent::BatchCommitStarted(evt) => evt.event_type(),
      RfcEvent::BatchCommitEnded(evt) => evt.event_type(),
    }
  }
}

impl WithEventContext for RfcEvent {
  fn with_event_context(&mut self, event_context: EventContext) {
    match self {
      RfcEvent::ContributionAdded(evt) => evt.event_context.replace(event_context),
      RfcEvent::APINamed(evt) => evt.event_context.replace(event_context),
      RfcEvent::GitStateSet(evt) => evt.event_context.replace(event_context),
      RfcEvent::BatchCommitStarted(evt) => evt.event_context.replace(event_context),
      RfcEvent::BatchCommitEnded(evt) => evt.event_context.replace(event_context),
    };
  }
}

impl Event for ContributionAdded {
  fn event_type(&self) -> &'static str {
    "ContributionAdded"
  }
}

impl Event for APINamed {
  fn event_type(&self) -> &'static str {
    "APINamed"
  }
}

impl Event for GitStateSet {
  fn event_type(&self) -> &'static str {
    "GitStateSet"
  }
}

impl Event for BatchCommitStarted {
  fn event_type(&self) -> &'static str {
    "BatchCommitStarted"
  }
}

impl Event for BatchCommitEnded {
  fn event_type(&self) -> &'static str {
    "BatchCommitEnded"
  }
}

impl From<BatchCommitStarted> for RfcEvent {
  fn from(event: BatchCommitStarted) -> Self {
    Self::BatchCommitStarted(event)
  }
}

impl From<BatchCommitEnded> for RfcEvent {
  fn from(event: BatchCommitEnded) -> Self {
    Self::BatchCommitEnded(event)
  }
}

impl From<ContributionAdded> for RfcEvent {
  fn from(event: ContributionAdded) -> Self {
    Self::ContributionAdded(event)
  }
}

// Conversion from commands
// ------------------------

impl From<RfcCommand> for RfcEvent {
  fn from(rfc_command: RfcCommand) -> Self {
    match rfc_command {
      RfcCommand::StartBatchCommit(command) => RfcEvent::from(BatchCommitStarted::from(command)),
      RfcCommand::EndBatchCommit(command) => RfcEvent::from(BatchCommitEnded::from(command)),
      RfcCommand::AddContribution(command) => RfcEvent::from(ContributionAdded::from(command)),
      _ => unimplemented!(
        "conversion from rfc command to rfc event not implemented for variant: {:?}",
        rfc_command
      ),
    }
  }
}

impl From<rfc_commands::StartBatchCommit> for BatchCommitStarted {
  fn from(command: rfc_commands::StartBatchCommit) -> Self {
    Self {
      batch_id: command.batch_id,
      parent_id: Some(command.parent_id),
      commit_message: command.commit_message,
      event_context: None,
    }
  }
}

impl From<rfc_commands::EndBatchCommit> for BatchCommitEnded {
  fn from(command: rfc_commands::EndBatchCommit) -> Self {
    Self {
      batch_id: command.batch_id,
      event_context: None,
    }
  }
}

impl From<rfc_commands::AddContribution> for ContributionAdded {
  fn from(command: rfc_commands::AddContribution) -> Self {
    Self {
      id: command.id,
      key: command.key,
      value: command.value,
      event_context: None,
    }
  }
}
