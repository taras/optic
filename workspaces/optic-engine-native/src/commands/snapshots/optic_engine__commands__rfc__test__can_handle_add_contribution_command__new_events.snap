---
source: workspaces/optic-engine-native/src/commands/rfc.rs
expression: new_events
---
[
    ContributionAdded(
        ContributionAdded {
            id: "path123.method",
            key: "purpose",
            value: "Name of Endpoint",
            event_context: None,
        },
    ),
]
